lib/trento_web/controllers/v1/tags_controller.ex

```diff
@@ -1,33 +1,85 @@
 defmodule TrentoWeb.V1.TagsController do
   use TrentoWeb, :controller
+  use OpenApiSpex.ControllerSpecs
 
   alias Trento.Tags
 
+  alias TrentoWeb.OpenApi.Schema
+
   action_fallback TrentoWeb.FallbackController
+  plug OpenApiSpex.Plug.CastAndValidate, json_render_error_v2: true
+
+  operation :add_tag,
+    summary: "Add tag",
+    tags: ["Tags"],
+    description: "Add a tag to a resource.",
+    parameters: [
+      id: [
+        in: :path,
+        required: true,
+        type: %OpenApiSpex.Schema{type: :string, format: :uuid}
+      ]
+    ],
+    request_body:
+      {"Tag", "application/json",
+       %OpenApiSpex.Schema{
+         type: :object,
+         properties: %{
+           value: %OpenApiSpex.Schema{type: :string}
+         }
+       }},
+    responses: [
+      accepted: "The tag has been added to the resource",
+      bad_request: Schema.BadRequest.response(),
+      unprocessable_entity: OpenApiSpex.JsonErrorResponse.response()
+    ]
 
-  @spec add_tag(Plug.Conn.t(), map) :: Plug.Conn.t()
   def add_tag(
-        %{assigns: %{resource_type: resource_type}} = conn,
         %{
-          "id" => id,
-          "value" => value
-        }
+          assigns: %{
+            resource_type: resource_type
+          }
+        } = conn,
+        %{id: id}
       ) do
+    %{value: value} = Map.get(conn, :body_params)
+
     case Tags.add_tag(value, id, resource_type) do
       {:ok, %Trento.Tag{value: value}} ->
         conn
         |> put_status(:created)
         |> json(%{value: value})
 
       {:error, errors} ->
-        {:error, {:bad_request, errors}}
+        {:error, {:unprocessable_entity, errors}}
     end
   end
 
-  @spec remove_tag(Plug.Conn.t(), map) :: Plug.Conn.t()
+  operation :remove_tag,
+    summary: "Remove tag",
+    tags: ["Tags"],
+    description: "Remove a tag from a resource.",
+    parameters: [
+      id: [
+        in: :path,
+        required: true,
+        type: %OpenApiSpex.Schema{type: :string, format: :uuid}
+      ],
+      value: [
+        in: :path,
+        required: true,
+        type: %OpenApiSpex.Schema{type: :string}
+      ]
+    ],
+    responses: [
+      accepted: "The tag has been removed from the resource",
```

Comment:
Currently this operation returns 204 No Content on success and possibly a 404.
Are we planning to update the operation based on this new api spec?

Reply:
as above, the 422 and 400 are gently offered by openapispex and phoenix.
202 is wrong and 404 is missing.